Mathlib/Combinatorics/HalesJewett.lean

```diff
@@ -368,4 +386,90 @@ theorem exists_mono_homothetic_copy {M κ : Type*} [AddCommMonoid M] (S : Finset
+@[ext]
+structure Subspace (η α ι : Type*) where
+  /-- The word representing a combinatorial subspace. `l.idxfun i = Sum.inr e` means that
+  `l x i = x e` for all `x` and `l.idxfun i = some a` means that `l x i = a` for all `x`. -/
+  idxFun : ι → α ⊕ η
+  /-- We require combinatorial subspaces to be nontrivial in the sense that `fun x ↦ l x i` is
+  `fun x ↦ x e` for at least one coordinate `i`. -/
+  proper : ∀ e, ∃ i, idxFun i = Sum.inr e
+
+namespace Subspace
+variable {η α ι κ : Type*} {l : Subspace η α ι} {x : η → α} {i : ι} {a : α} {e : η}
+
+/-- The combinatorial subspace corresponding to the identity embedding `(ι → α) → (ι → α)`. -/
+instance : Inhabited (Subspace ι α ι) := ⟨⟨Sum.inr, fun i ↦ ⟨i, rfl⟩⟩⟩
+
+instance instCoeFun : CoeFun (Subspace η α ι) (fun _ ↦ (η → α) → ι → α) :=
+  ⟨fun l x i ↦ (l.idxFun i).elim id x⟩
+
+-- Note: This is not made a `FunLike` instance to avoid having two syntactically different coercions
+lemma coe_injective [Nontrivial α] : Injective ((⇑) : Subspace η α ι → (η → α) → ι → α) := by
+  rintro l m hlm
+  ext i
+  simp only [funext_iff] at hlm
+  cases hl : idxFun l i with
+  | inl a =>
+    obtain ⟨b, hba⟩ := exists_ne a
+    cases hm : idxFun m i <;> simpa [hl, hm, hba.symm] using hlm (const _ b) i
+  | inr e =>
+    cases hm : idxFun m i with
+    | inl a =>
+      obtain ⟨b, hba⟩ := exists_ne a
+      simpa [hl, hm, hba] using hlm (const _ b) i
+    | inr f =>
+      obtain ⟨a, b, hab⟩ := exists_pair_ne α
+      simp only [Sum.inr.injEq]
+      by_contra! hef
+      simpa [hl, hm, hef, hab] using hlm (Function.update (const _ a) f b) i
+
+lemma apply_def (l : Subspace η α ι) (x : η → α) (i : ι) : l x i = (l.idxFun i).elim id x := rfl
+lemma apply_inl (h : l.idxFun i = Sum.inl a) : l x i = a := by simp [apply_def, h]
+lemma apply_inr (h : l.idxFun i = Sum.inr e) : l x i = x e := by simp [apply_def, h]
+
+/-- Given a coloring `C` of `ι → α` and a combinatorial subspace `l` of `ι → α`, `l.IsMono C`
+means that `l` is monochromatic with regard to `C`. -/
+def IsMono (C : (ι → α) → κ) (l : Subspace η α ι) : Prop := ∃ c, ∀ x, C (l x) = c
+
+/-- The **extended Hales-Jewett theorem**: For any finite types `η`, `α` and `κ`, there exists a
+finite type `ι` such that whenever the hypercube `ι → α` is `κ`-colored, there is a monochromatic
+combinatorial subspace of dimension `η`. -/
+theorem exists_mono_in_high_dimension (α κ η) [Fintype α] [Fintype κ] [Fintype η] :
+    ∃ (ι : Type) (_ : Fintype ι), ∀ C : (ι → α) → κ, ∃ l : Subspace η α ι, l.IsMono C := by
+  obtain ⟨ι, _, hι⟩ := Line.exists_mono_in_high_dimension (η → α) κ
+  refine ⟨ι × Shrink η, inferInstance, fun C ↦ ?_⟩
+  obtain ⟨l, c, lC⟩ := hι fun x ↦ C <| fun (i, e) ↦ x i <| (equivShrink.{0} η).symm e
+  refine ⟨⟨fun p ↦ (l.idxFun p.fst).elim (Sum.inr <| (equivShrink.{0} η).symm p.snd)
+    fun x ↦ Sum.inl <| x <| (equivShrink.{0} η).symm p.snd, fun e ↦ ?_⟩, c, fun xs ↦ ?_⟩
```

Comment:
Is there a `Line.toSubspace` definition hiding here?

Reply:
There was, but damn that was not easy.